runner/docker-shim.sh

```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+
+set -Eeuo pipefail
+
+if [[ ${ARC_DOCKER_MTU_PROPAGATION:-false} == true ]] &&
+  (($# >= 2)) && [[ $1 == network && $2 == create ]] &&
+  mtu=$(/usr/bin/docker network inspect bridge --format '{{index .Options "com.docker.network.driver.mtu"}}' 2>&-); then
```

Comment:
I'm afraid of using `2>&-` as we encountered #1560 due to our lack of understanding of how `FD>&-` works.
Can we perhaps use `2>/dev/null` as a safe alternative?

Reply:
Actually we do understand what it means to close a file descriptor, and there's no need to be afraid of it. Vigilant, yes, testing properly, yes, but no need to be afraid. Goroutines suffer from races and deadlocks, we still use them, and are not switching to Rust because of it. 😉 That said, the difference is marginal in this case. Swallowing stderr is often the bigger issue, as it might hide issues. Hence, we could also just let it inherit stderr and print.

Reply:
Thanks! But did we already know why `1>&-` caused the problem in some envs and not in other envs? That's my concern.

It does look like a "correct" way to write a bash script, but I thought we are yet to see any proof or data point about why `1>&-` can be considered reliable, right? 🤔

Reply:
> we do understand what it means to close a file descriptor

To be extra clear, my point is if closing it to suppress some output is the right way to do it, as it can break like in #1560. `2>/dev/null` doesn't close it(right?) so it won't result in issues like #1560. Why use `1>&-` then..?

Reply:
Yes, I explained it in https://github.com/actions-runner-controller/actions-runner-controller/pull/1560#issuecomment-1170961218. Using `FD>&-` is the same as `File.Close` in Go. We are essentially closing the file descriptor so that all writes to it simply never happen, no bits are ever being transferred. Using `FD>/dev/null` means that all bits are still transferred to the device. In this case a device that does nothing with the bits, but they still need to be transferred.

The problem with `FD>&-` arises if the caller does never check if the file descriptor (or file handle) is actually writable, and simply writes. This is obviously going to fail, since the file is not writable, it is closed.

Many programs are expecting FD1 to always be writable, no matter what. That is what we saw in the other issue. However, the same is not true for any FD greater one. Here is a little more on this in the context of C specifically: https://stackoverflow.com/a/4973065/1251219

I am totally fine with using `>/dev/null` everywhere for safety reasons. It is just important to me that everyone understands the difference. This really has nothing to do with Bash, it is just normal programming in relation to how to deal with file handles, and the same in every programming language where we have full control over file handles.

Reply:
https://www.kernel.org/doc/html/latest/block/null_blk.html according to this `/dev/null` does not actually execute any read/write operations. Maybe there really is no difference anymore in modern Linux.

Reply:
First of thanks a lot for sharing your insights and knowledge. I learned something!

> Yes, I explained it in https://github.com/actions-runner-controller/actions-runner-controller/pull/1560#issuecomment-1170961218

Ahh, thanks! Sorry, I somehow missed it.

> The problem with FD>&- arises if the caller does never check if the file descriptor (or file handle) is actually writable, and simply writes. This is obviously going to fail, since the file is not writable, it is closed.

This makes sense a lot! Thanks again for clarifying.

Does it basically mean that when using `$FD>&-`, we need to be sure if the target program checks the availability of the FD before writing to it? 

In this specific case, the use of `2>&-` is fine as long as `docker network inspect bridge` either checks for the availability of the fd 2(stderr) before writing, or it handles stderr write errors gracefully, right?

Reply:
I've updated the code to use `2>/dev/null` redirection instead. According to @Fleshgrinder redirecting to `/dev/null` does actually execute writes so there's no difference.

Reply:
@tiagoblackcode way to go.

@mumoshu yes, that is what one need to know about the program before file handles can be closed. Normal programs can deal with stderr being closed, but not stdin and stdout.